pkg/util/strategicpatch/fields.go

```diff
@@ -0,0 +1,469 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// NOTE: The below is taken from the Go standard library to enable us to find
+// the field of a struct that a given JSON key maps to.
+//
+// Copyright 2013 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+package strategicpatch
```

Comment:
The functions are just copied. The issue is that these are all private functions listed from the encoding/json package. I don't know how to package this well because if you put them in another package, you have to change the functions to be public, which I don't know is the best idea. I had to do the exact same thing for the ghodss/yaml package, see https://github.com/ghodss/yaml/blob/master/fields.go.

The problem is that Go's rules for how it finds a field based on a struct definition and a JSON field name is an interface that Go cannot break (at least in 1.0). However, they do not expose any functions or packages to actually let you execute this lookup yourself. So I just copy/pasted the same functions that encoding/json uses to do this calculation. The functions in this file work because they are tested in the go package, and they should never be modified in here.